diff -Naur xonsh-0.9.13_orig/xonsh/__amalgam__.py xonsh-0.9.13/xonsh/__amalgam__.py
--- xonsh-0.9.13_orig/xonsh/__amalgam__.py	2020-01-15 18:34:16.891000000 +0100
+++ xonsh-0.9.13/xonsh/__amalgam__.py	2020-01-15 18:36:21.998000000 +0100
@@ -21251,7 +21251,7 @@
 
 
 def xexec(args, stdin=None):
-    """exec [-h|--help] command [args...]
+    """exec [-h|--help] [-cl] [-a name] command [args...]
 
     exec (also aliased as xexec) uses the os.execvpe() function to
     replace the xonsh process with the specified program. This provides
@@ -21261,6 +21261,13 @@
         bash $
 
     The '-h' and '--help' options print this message and exit.
+    If the '-l' option is supplied, the shell places a dash at the
+    beginning of the zeroth argument passed to command to simulate login
+    shell.
+    The '-c' option causes command to be executed with an empty environment.
+    If '-a' is supplied, the shell passes name as the zeroth argument
+    to the executed command.
+
 
     Notes
     -----
@@ -21274,18 +21281,39 @@
     """
     if len(args) == 0:
         return (None, "xonsh: exec: no args specified\n", 1)
-    elif args[0] == "-h" or args[0] == "--help":
+
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument('-h', '--help', action='store_true')
+    parser.add_argument('-l', dest='login', action='store_true')
+    parser.add_argument('-c', dest='clean', action='store_true')
+    parser.add_argument('-a', dest='name', nargs='?')
+    parser.add_argument('command', nargs=argparse.REMAINDER)
+    args = parser.parse_args(args)
+
+    if args.help:
         return inspect.getdoc(xexec)
-    else:
+
+    if len(args.command) == 0:
+        return (None, "xonsh: exec: no command specified\n", 1)
+
+    command = args.command[0]
+    if args.name is not None:
+        args.command[0] = args.name
+    if args.login:
+        args.command[0] = '-{}'.format(args.command[0])
+
+    denv = {}
+    if not args.clean:
         denv = builtins.__xonsh__.env.detype()
-        try:
-            os.execvpe(args[0], args, denv)
-        except FileNotFoundError as e:
-            return (
-                None,
-                "xonsh: exec: file not found: {}: {}" "\n".format(e.args[1], args[0]),
-                1,
-            )
+
+    try:
+        os.execvpe(command, args.command, denv)
+    except FileNotFoundError as e:
+        return (
+            None,
+            "xonsh: exec: file not found: {}: {}" "\n".format(e.args[1], args[0]),
+            1,
+        )
 
 
 class AWitchAWitch(argparse.Action):
@@ -23908,7 +23936,7 @@
     env = builtins.__xonsh__.env
     rc = shell_kwargs.get("rc", None)
     rc = env.get("XONSHRC") if rc is None else rc
-    if args.mode != XonshMode.interactive and not args.force_interactive:
+    if args.mode != XonshMode.interactive and not args.force_interactive and not args.login:
         #  Don't load xonshrc if not interactive shell
         rc = None
     events.on_pre_rc.fire()
@@ -23945,7 +23973,8 @@
         "cacheall": args.cacheall,
         "ctx": builtins.__xonsh__.ctx,
     }
-    if args.login:
+    if args.login or sys.argv[0].startswith('-'):
+        args.login = True
         shell_kwargs["login"] = True
     if args.norc:
         shell_kwargs["rc"] = ()
